File: common/src/main/java/com/terrasect/common/definition/NoiseConstraints.java
package com.terrasect.common.definition;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import java.util.function.Consumer;

public record NoiseConstraints(
    Map<String, NoiseTransform> noises,
    Map<String, NoiseTransform> densityFunctions,
    boolean clearsParent) {

  public NoiseConstraints {
    if (noises == null) noises = Collections.emptyMap();
    if (densityFunctions == null) densityFunctions = Collections.emptyMap();
    noises = Map.copyOf(noises);
    densityFunctions = Map.copyOf(densityFunctions);
  }

  public boolean hasAnyConstraints() {
    return !noises.isEmpty() || !densityFunctions.isEmpty();
  }

  public static NoiseConstraints empty() {
    return new NoiseConstraints(Collections.emptyMap(), Collections.emptyMap(), false);
  }

  public NoiseConstraints resolveWithParent(NoiseConstraints parent) {
    if (clearsParent) return this;
    if (parent == null) return this;
    if (!parent.hasAnyConstraints()) return this;

    var mergedNoises = new LinkedHashMap<String, NoiseTransform>(parent.noises);
    mergedNoises.putAll(noises);

    var mergedDensity = new LinkedHashMap<String, NoiseTransform>(parent.densityFunctions);
    mergedDensity.putAll(densityFunctions);

    return new NoiseConstraints(mergedNoises, mergedDensity, false);
  }

  public static Builder builder() {
    return new Builder();
  }

  public static final class Builder {
    private boolean clearParent = false;
    private final Map<String, NoiseTransform> noises = new LinkedHashMap<>();
    private final Map<String, NoiseTransform> densityFunctions = new LinkedHashMap<>();

    public Builder clearParent() {
      clearParent = true;
      return this;
    }

    public Builder noise(String noiseKey, Consumer<NoiseTransform.Builder> consumer) {
      Objects.requireNonNull(noiseKey, "noiseKey");
      Objects.requireNonNull(consumer, "consumer");

      var builder = NoiseTransform.builder().copyFrom(noises.get(noiseKey));
      consumer.accept(builder);
      var transform = builder.build();
      if (transform.isEmpty()) {
        noises.remove(noiseKey);
      } else {
        noises.put(noiseKey, transform);
      }
      return this;
    }

    public Builder densityFunction(
        String densityFunctionKey, Consumer<NoiseTransform.Builder> consumer) {
      Objects.requireNonNull(densityFunctionKey, "densityFunctionKey");
      Objects.requireNonNull(consumer, "consumer");

      var builder = NoiseTransform.builder().copyFrom(densityFunctions.get(densityFunctionKey));
      consumer.accept(builder);
      var transform = builder.build();
      if (transform.isEmpty()) {
        densityFunctions.remove(densityFunctionKey);
      } else {
        densityFunctions.put(densityFunctionKey, transform);
      }
      return this;
    }

    public Builder copyFrom(NoiseConstraints constraints) {
      if (constraints == null) return this;
      clearParent = constraints.clearsParent();
      noises.clear();
      densityFunctions.clear();
      noises.putAll(constraints.noises());
      densityFunctions.putAll(constraints.densityFunctions());
      return this;
    }

    public NoiseConstraints build() {
      var builtNoises = new LinkedHashMap<String, NoiseTransform>(noises);
      var builtDensity = new LinkedHashMap<String, NoiseTransform>(densityFunctions);
      return new NoiseConstraints(builtNoises, builtDensity, clearParent);
    }
  }
}

File: common/src/main/kotlin/terrasect/definition/NoiseConstraints.kt
package terrasect.definition

import terrasect.helpers.NoiseTransform

class NoiseConstraints(
  val noises: Map<String, NoiseTransform>,
  val densityFunctions: Map<String, NoiseTransform>,
  val blendWidth: Float = DEFAULT_BLEND_WIDTH,
) {
  companion object {
    const val DEFAULT_BLEND_WIDTH: Float = 32f

    fun builder() = Builder()
  }

  fun hasAnyConstraints(): Boolean = noises.isNotEmpty() || densityFunctions.isNotEmpty()

  class Builder {
    private val noises = mutableMapOf<String, NoiseTransform>()
    private val densityFunctions = mutableMapOf<String, NoiseTransform>()
    private var blendWidth: Float = DEFAULT_BLEND_WIDTH
    private var blendWidthExplicit: Boolean = false

    fun noise(name: String, transform: NoiseTransform) = apply { noises[name] = transform }

    fun noise(name: String, consumer: (NoiseTransform.Builder) -> Unit) = apply {
      val transform = NoiseTransform.builder().apply(consumer).build()
      noises[name] = transform
    }

    fun densityFunction(name: String, transform: NoiseTransform) = apply {
      densityFunctions[name] = transform
    }

    fun densityFunction(name: String, consumer: (NoiseTransform.Builder) -> Unit) = apply {
      val transform = NoiseTransform.builder().apply(consumer).build()
      densityFunctions[name] = transform
    }

    fun blendWidth(width: Float) = apply {
      this.blendWidth = width
      this.blendWidthExplicit = true
    }

    fun inheritParent(parent: Builder) = apply {
      for ((name, transform) in parent.noises) {
        this.noises.putIfAbsent(name, transform)
      }
      for ((name, transform) in parent.densityFunctions) {
        this.densityFunctions.putIfAbsent(name, transform)
      }
      if (!blendWidthExplicit) {
        this.blendWidth = parent.blendWidth
      }
    }

    fun build(): NoiseConstraints {
      return NoiseConstraints(noises.toMap(), densityFunctions.toMap(), blendWidth)
    }
  }
}
